### PR TITLE
#9911: Add structure and migrate 20 composite unary ops

### PR DESCRIPTION
**Pull request by bharane-ab**
### Ticket
Unary Backward Tracking Issue : #9911  [Master issue : #9322  ]

list of unary composite op migrated:
- [x] acosh
- [x] asinh
- [x] atanh
- [x] cbrt
- [x] cosh
- [x] digamma
- [x] lgamma
- [x] log1p
- [x] mish
- [x] multigammaln
- [x] sinh
- [x] softsign
- [x] swish
- [x] tanhshrink
- [x] trunc
- [x] var_hw
- [x] std_hw
- [x] normalize_hw
- [x] deg2rad
- [x] rad2deg

### Checklist
- [x] All post commit [link](https://github.com/tenstorrent/tt-metal/actions/runs/9908732011)
- [x]  post commit c++ [link](https://github.com/tenstorrent/tt-metal/actions/runs/9905424777)
- [x] post commit model test [link](https://github.com/tenstorrent/tt-metal/actions/runs/9905432827)
- [x] T3k [link](https://github.com/tenstorrent/tt-metal/actions/runs/9905446618)

### CODE REVIEW

**Review by eyonland — APPROVED** [review]
It looks fine to me.

Let's get the ttnn unit tests updated instead of leaving them in eager.

**Comment by eyonland** [comment]
Please add optional output tensors and qid of course.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp [hunk 1]
```diff
@@ -86,6 +87,36 @@ struct Power{
         return detail::power_fp(DefaultQueueId, input_tensor, exponent, memory_config, optional_output_tensor);
     }
 };
+template <UnaryCompositeOpType unary_comp_op_type>
+struct Unary_composite_ops
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp:91]:
```suggestion
struct ExecuteUnaryCompositeOp
```

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp [hunk 2]
```diff
@@ -86,6 +87,36 @@ struct Power{
...
+
+    static ttnn::Tensor execute_on_worker_thread(
+        const Tensor& input_tensor,
+        int value_1,
+        const std::optional<MemoryConfig>& memory_config = std::nullopt)
+        {
+            auto op_type = UnaryCompositeFunction::get_function_type3(unary_comp_op_type);
+            return op_type(input_tensor, value_1, memory_config);
+        }
+};
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp:119]: @eyonland , adding queue_id / optional_output_tensor separately, correct?

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.hpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    SWISH,
+    TANHSHRINK,
+    TRIL,
+    TRIU,
+};
+struct UnaryCompositeFunction
+{
+    static std::function<Tensor(const Tensor&, const std::optional<MemoryConfig>&)> get_function_type1(UnaryCompositeOpType OpType);
+    static std::function<Tensor(const Tensor&, float, float, const std::optional<MemoryConfig>&)> get_function_type2(UnaryCompositeOpType OpType);
+    static std::function<Tensor(const Tensor&, int, const std::optional<MemoryConfig>&)> get_function_type3(UnaryCompositeOpType OpType);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.hpp:38]: please coordinate with @eyonland and @KalaivaniMCW if this should be fixed now or will have a follow up fix.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp (hunk 2 above)
Comment (eyonland) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp:119]: Yea this is part of M2 as defined in [the status ](https://docs.google.com/document/d/1B74vJPjpUCjM3c3K3QOShQznDUNCdBAlcu5J3V2671Y/edit)

**Review by eyonland — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_unary_ops_ttnn.py [hunk 4]
```diff
@@ -6,6 +6,7 @@
 import pytest
 import ttnn
 from tests.tt_eager.python_api_testing.unit_testing.backward_ops.utility_funcs import data_gen_with_range, compare_pcc
```
Comment (KalaivaniMCW) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_unary_ops_ttnn.py:8]: this file will go under tests/ttnn/unit_tests/operations/composite
and use tests/ttnn/unit_tests/operations/backward/utility_funcs.py for import

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp [hunk 5]
```diff
@@ -0,0 +1,388 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+
+#include <functional>
+#include <optional>
+#include <iostream>
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp:8]: can be used during testing.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp [hunk 6]
```diff
@@ -86,6 +87,36 @@ struct Power{
...
 
 // re-implement tt_eager composite unary op => ttnn composite unary ops.
 Tensor deg2rad(uint8_t queue_id, const Tensor& input_tensor, const std::optional<MemoryConfig>& memory_config = std::nullopt, const std::optional<Tensor>& optional_output_tensor = std::nullopt) {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp:122]: can be moved to unary_composite_op.cpp to match existing struct ExecuteUnaryCompositeOp

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp [hunk 7]
```diff
@@ -249,27 +237,29 @@ constexpr auto deg2rad = ttnn::register_operation("ttnn::deg2rad", TO_LAMBDA_WIT
 constexpr auto rad2deg = ttnn::register_operation("ttnn::rad2deg", TO_LAMBDA_WITH_RESHAPE(ttnn::operations::unary::rad2deg));
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_composite.hpp:237]: can be implemented like the ops below.  can be moved to unary_composite_op.cpp to match existing struct ExecuteUnaryCompositeOp

File: tests/ttnn/unit_tests/operations/test_math.py [hunk 8]
```diff
@@ -26,6 +26,8 @@ def run_math_unary_test(device, h, w, ttnn_function, torch_function, pcc=0.9999)
     input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
     output_tensor = ttnn_function(input_tensor)
     output_tensor = ttnn.to_torch(output_tensor)
+    if ttnn_function == ttnn.cbrt:
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_math.py:29]: Why is it the right thing to add an if for a specific operation here?

File: tests/ttnn/unit_tests/operations/test_unary.py [hunk 9]
```diff
@@ -23,6 +23,8 @@ def run_unary_test(device, h, w, ttnn_function, torch_function, pcc=0.9999):
     output_tensor = ttnn.to_layout(output_tensor, ttnn.ROW_MAJOR_LAYOUT)
     output_tensor = ttnn.from_device(output_tensor)
     output_tensor = ttnn.to_torch(output_tensor)
+    if ttnn_function == ttnn.acosh:
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_unary.py:26]: Same concern

File: tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py [hunk 10]
```diff
@@ -465,7 +465,7 @@ def eltwise_relu_max(
 @setup_host_and_device
 def std_hw(x, *args, device, dtype, layout, input_mem_config, output_mem_config, **kwargs):
     t0 = setup_tt_tensor(x, device, layout[0], input_mem_config[0], dtype[0])
-    t1 = ttl.tensor.std_hw(t0, output_mem_config=output_mem_config)
+    t1 = ttnn.std_hw(t0)
```
Comment (eyonland) [inline tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py:468]: Yet another example of a function that deviates from PyTorch.  Not part of this PR obviously but we need to get these fixed soon.

**Review by eyonland — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

File: tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py (hunk 10 above)
Comment (bharane-ab) [reply tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py:468]: Yes this will be handled when we work on M6.
